File: petclinic-java-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/java/dev/pollito/petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts/user/UserService.java
package dev.pollito.petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts.user;

import dev.pollito.petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts.events.BeforeDeleteOwner;
import dev.pollito.petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.ReferencedException;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;

@Service
public class UserService {

    private final UserRepository userRepository;

    public UserService(final UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @EventListener(BeforeDeleteOwner.class)
    public void on(final BeforeDeleteOwner event) {
        final ReferencedException referencedException = new ReferencedException();
        final User ownerUser = userRepository.findFirstByOwnerId(event.getId());
        if (ownerUser != null) {
            referencedException.setKey("owner.user.owner.referenced");
            referencedException.addParam(ownerUser.getUsername());
            throw referencedException;
        }
    }

}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/user/UserService.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.user

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.events.BeforeDeleteOwner
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.ReferencedException
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Service


@Service
class UserService(
    private val userRepository: UserRepository
) {

    @EventListener(BeforeDeleteOwner::class)
    fun on(event: BeforeDeleteOwner) {
        val referencedException = ReferencedException()
        val ownerUser = userRepository.findFirstByOwnerId(event.id)
        if (ownerUser != null) {
            referencedException.key = "owner.user.owner.referenced"
            referencedException.addParam(ownerUser.username)
            throw referencedException
        }
    }

}
